specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/AttachmentsUpload.tsx

```diff
@@ -0,0 +1,428 @@
+import React from 'react';
+
+import { usePromise } from '../../hooks/useAsyncState';
+import { attachmentsText } from '../../localization/attachments';
+import { commonText } from '../../localization/common';
+import { wbText } from '../../localization/workbench';
+import { ajax } from '../../utils/ajax';
+import type { RA } from '../../utils/types';
+import { filterArray } from '../../utils/types';
+import { formatTime } from '../../utils/utils';
+import { Progress } from '../Atoms';
+import { Button } from '../Atoms/Button';
+import { dialogIcons } from '../Atoms/Icons';
+import {
+  attachmentSettingsPromise,
+  uploadFile,
+} from '../Attachments/attachments';
+import { serializeResource } from '../DataModel/helpers';
+import type { SerializedResource } from '../DataModel/helperTypes';
+import type { SpecifyResource } from '../DataModel/legacyTypes';
+import { schema, strictGetModel } from '../DataModel/schema';
+import type { Attachment, Tables } from '../DataModel/types';
+import { Dialog, LoadingScreen } from '../Molecules/Dialog';
+import type { AttachmentUploadSpec, EagerDataSet } from './Import';
+import { PerformAttachmentTask } from './PerformAttachmentTask';
+import type {
+  AttachmentStatus,
+  AttachmentWorkStateProps,
+  PartialUploadableFileSpec,
+  UploadAttachmentSpec,
+} from './types';
+import {
+  getAttachmentsFromResource,
+  resolveAttachmentRecord,
+  validateAttachmentFiles,
+} from './utils';
+
+async function prepareForUpload(
+  dataSet: EagerDataSet,
+  baseTableName: keyof Tables
+): Promise<RA<PartialUploadableFileSpec>> {
+  const validatedFiles = await validateAttachmentFiles(
+    dataSet.uploadableFiles,
+    dataSet.uploadSpec as AttachmentUploadSpec,
+    // If user validated before, and chose disambiguation, need to preserve it
+    true
+  );
+  const mappedUpload = await Promise.all(
+    validatedFiles.map(async (uploadable) =>
+      uploadFileWrapped(uploadable, baseTableName, undefined, true)
+    )
+  );
+  const fileNamesToTokenize = filterArray(
+    mappedUpload.map((uploadable) =>
+      uploadable.status?.type === 'matched' &&
+      uploadable.uploadTokenSpec === undefined
+        ? uploadable.file.name
+        : undefined
+    )
+  );
+  if (fileNamesToTokenize.length === 0) return mappedUpload;
+  return ajax<RA<UploadAttachmentSpec>>('/attachment_gw/get_upload_params/', {
+    method: 'POST',
+    headers: { Accept: 'application/json' },
+    body: {
+      filenames: fileNamesToTokenize,
+    },
+  }).then(({ data }) => {
+    if (fileNamesToTokenize.length !== data.length) {
+      // Throwing an error for development testing. Hasn't happened yet.
+      throw new Error(
```

Comment:
don't need to change anything here, but just a FYI, that we have a `softError` function - it throws an error in development, but only console errors in production